### ISSUE TITLE
[Java] NoClassDefFoundError

### ISSUE THREAD

**Issue by MPNull** [codenvy/codenvy#693]
> Exception in thread "main" java.lang.NoClassDefFoundError: org/json/JSONObject

I am wondering why it is so hard to make a Maven projects work fine with external JAR libraries. I've added the needed information to pom.xml. The project will download all of the required dependencies and also IDE doesn't mark the imported packages or syntax markups as invalid.

```
    <dependencies>
        <dependency>
            <groupId>junit</groupId>
            <artifactId>junit</artifactId>
            <version>3.8.1</version>
            <scope>test</scope>
        </dependency>
        <!-- https://mvnrepository.com/artifact/org.json/json -->
        <dependency>
            <groupId>org.json</groupId>
            <artifactId>json</artifactId>
            <version>20160810</version>
        </dependency>
        <!-- https://mvnrepository.com/artifact/com.googlecode.json-simple/json-simple -->
        <dependency>
            <groupId>com.googlecode.json-simple</groupId>
            <artifactId>json-simple</artifactId>
            <version>1.1.1</version>
        </dependency>
    </dependencies>
```

I am trying to use org.json and org.json.simple, but it doesn't matter how ever I am trying to import it. I've read around the internet about this error, but it doesn't make any sense for me. I was so excited about it that I don't need to configure something "hard" what I don't understand, but now I have to. Normally I use Eclipse on PC, but when I'm not at home it would be awesome to code something at somewhere without installing the stuff on the PC I'm using.

> In Codenvy, build and run processes take place on different nodes and different environments. This project produces /repo with quite a few jars and a startup script that is impossible to inject into a Docker container (Codenvy runners are Docker based). Thus, I recommend performing build and run in the same environment, i.e. in the runtime. What you need to do is create a new runner (the button is on the runner panel), install Maven there, perform build and execute startup script. You will also have to unbind the service from localhost -kalrsson from StackOverflow

So basically what can I do? I really would like to use Codenvy of course.

**Comment by TylerJewell** [comment]
@ddementieva @gazarenkov @eivantsov @JamesDrummond - can one of you comment.  It should be as simple as just adding the dependency to the pom.xml.  Is there something that our systems are doing to block lookup against the nexus?

**Comment by MPNull** [comment]
@TylerJewell Well. I added the dependencies to pom.xml and it downloads them successfully. They're located in "External Libraries" as well as "junit" and other libraries. I'm super new with Codenvy and that's probably the reason why it feels so hard. Normally I use Eclipse and I just add the specific .jar files from project properties. After adding them from project properties they're available for use by importing them to any .java file. I've never been using a Maven project, but it was pretty easy to setup using a Codenvy. All I understood about Maven is that the Maven provides the repositories for the external libraries etc. So shouldn't they be usable after adding them as dependencies in pom.xml? I was browsing a Maven Repository with a browser and I copied the dependency "snippet" from it and I pasted it under another dependency in <dependencies>

**Comment by TylerJewell** [comment]
In a maven project - yes, it should just be as simple as adding them to your pom.xml and they  are immediately available from your project.  If it's not working that way, then I am worried something isn't operating properly.  Codenvy is authored using maven itself, so our engineers are very maven-centric and this stuff should just work properly.

As for the other way - you are thinking of a project type which is javac.  If you set the project type to javac, then you can upload the JARs to anywhere in your project tree, and then in Project -> Configuration - you can set the source and JAR path location for it to use, and then you get intellisense that way.

@JamesDrummond - can you also do a double check on our docs + add a link here so that this is helpful and clear for users.  I think this is twice in the same week this particular concern has shown up - so something isn't deadly intuitive.

**Comment by MPNull** [comment]
```
[default] Exception in thread "main" java.lang.NoClassDefFoundError: org/json/JSONObject
[default] 	at org.auctioneer.Auctioneer.main(Auctioneer.java:7)
[default] Caused by: java.lang.ClassNotFoundException: org.json.JSONObject
[default] 	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
[default] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
[default] 	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
[default] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
[default] 	... 1 more
```

**Comment by MPNull** [comment]
@TylerJewell Can I change the project type to javac on the run or do I have to create a new project?

**Comment by TylerJewell** [comment]
Neither - you just go into the Project menu and change the project type in there permanently.  It doesn't change any of your files, it just changes the meta data and changes how the IDE behaves.  You can switch back and forth.

**Comment by MPNull** [comment]
@TylerJewell By the way. I successfully changed the project to Java instead of Maven. In list of .jar files it shows that "junit", "json-xx" and "json-simple-xx" so I guess they was loaded up correctly when Maven was active, but I don't still understand why the project was not able to reach the libraries when I ran my code.

I followed this tutorial: http://www.tutorialspoint.com/json/json_java_example.htm
I didn't add anything else than I defined JSONObject like you see above. Libraries were not usable even if I added them to pom.xml + imported them in my Java file.

But anyways! I'm gonna try to add my local jars after 30 minutes. I have couple of things to do before continuing with Codenvy. Let's see what happens. Hope they're available then. I'll return back to this post then with a results. I already created a folder "lib" to my project base folder and defined it when I changed the project from Maven to Java.

Thanks for the help already!

**Comment by TylerJewell** [comment]
In about 4 hours, we'll have a bunch more engineers online and they can also start to help out, too.

**Comment by MPNull** [comment]
@TylerJewell 

1. I changed Maven project to Java there was option to define a location of "libraries" I pointed it to "lib" folder.
2. I uploaded the .jar files to folder called "lib" I just created. 
3. I changed classpath settings and chosen both of the libraries (JSON Simple & JSON)
4. I ran the program using same runner that came with the project "auctioneer:run"
5. Same error.

I am not sure should I create a new runner or something? It still uses some Maven commands when running the application.

In Google with keyword about the error: "java.lang.NoClassDefFoundError: org/json/JSONObject" I found following: http://stackoverflow.com/questions/27504508/java-lang-noclassdeffounderror-org-json-jsonobject

They still try to solve the issue by "adding jars" so I think the issue is related to something that the jars cannot be loaded or something. My defination should be 100% fine, because it just creates a new instance of JSONObject and shouldn't cause any errors. I also checked out that the "org.json" contains the JSONObject just to be sure there's no mistakes. "org.json.JSONObject" was available.

Problem can be me or problem can be anywhere. I guess it's just me, but if any of you feel hopeless with me I can provide my workspace details or just check it out since I use this GitHub account to login into Codenvy. **There is nothing important so everything can be deleted, modified or whatever without any kind of questions.**

**Comment by ddementieva** [comment]
@MPNull I've just added the following dependencies to the Spring sample project, added import and everything worked OK for me:  
```
<!-- https://mvnrepository.com/artifact/org.json/json -->
        <dependency>
            <groupId>org.json</groupId>
            <artifactId>json</artifactId>
            <version>20160810</version>
        </dependency>
        <!-- https://mvnrepository.com/artifact/com.googlecode.json-simple/json-simple -->
        <dependency>
            <groupId>com.googlecode.json-simple</groupId>
            <artifactId>json-simple</artifactId>
            <version>1.1.1</version>
        </dependency>
```

![screenshot](https://cloud.githubusercontent.com/assets/6860395/18306185/51946d5c-74f3-11e6-93d5-4c506e87362d.png)

Is it possible for you to share an entire project, e.g. as a git repository?

**Comment by MPNull** [comment]
Downloaded it: https://www.dropbox.com/s/ogjph2o8h4scfkg/auctioneer.zip?dl=0
I found sharing option, but it wants me to specify e-mail addresses to share with.

**Comment by MPNull** [comment]
People are talking on StackOverflow that the compiled JAR file cannot read the dependencies.

**Comment by TylerJewell** [comment]
@MPNull - can you crosslink the stackoverflow issue here, so that we can track them all and get closure with full resolution?

**Comment by MPNull** [comment]
PS. I've been trying to Google a solution to this issue like +6 hours and I'm pretty much starting lose my hope about finding a solution since I am not pretty familiar with Java at all. There's something I've fucked up I guess. It's weird if it works for someone else. I've tried to make a new workspace etc. None of them solved the issue. Building the stuff works fine, but when I run the code with "JSONObject jsonObject = new JSONObject();" it says me the error about "NoClassDef..."

I just tested it today on Eclipse. I added external JAR files via settings and everything worked up fine. Even I were able to fetch a data from JSON file.

**Comment by MPNull** [comment]
http://stackoverflow.com/questions/30618619/runtime-error-in-codenvy-java-execution

Solution is here. I don't know how to make it work anyways, but it would be awesome if it would work without custom changes otherwise I have to remember this all forever in order to make workspaces work properly with dependencies.

**Comment by TylerJewell** [comment]
I think the discussion online is for our legacy platform - the one that you are using is a different system, completely different code base - we put the new one out earlier this year.

**Comment by MPNull** [comment]
Mmh. But it's weird anyways that I can't get dependencies work. IMO I set them up correctly since they appear to external libraries.

**Comment by ddementieva** [comment]
@MPNull It seems you need to add Maven dependency plugin to copy your dependencies as jars to the lib folder, specified in your manifest. 

Simply add the following to the pom.xml, build and run your project:
```
<plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-dependency-plugin</artifactId>
        <executions>
            <execution>
                <id>copy-dependencies</id>
                <phase>install</phase>
                <goals>
                    <goal>copy-dependencies</goal>
                </goals>
                <configuration>
                    <outputDirectory>${project.build.directory}/lib</outputDirectory>
                    <overWriteReleases>false</overWriteReleases>
                    <overWriteSnapshots>false</overWriteSnapshots>
                    <overWriteIfNewer>true</overWriteIfNewer>
                </configuration>
            </execution>
        </executions>
    </plugin> 
```

**Comment by MPNull** [comment]
@ddementieva awesome, awesome, awesome! It was the solution! Luv U <3

That's one of the reasons why I didn't never treat a Codenvy about the problem. I were pretty sure that there's something I've fuck'd up myself. Thank you a lot.
